### PR TITLE
✨ answer progress bar + votes missing

### PR DESCRIPTION
**Pull request by staff0rd**
https://mobitroll.atlassian.net/browse/FRES-1405

### CODE REVIEW

File: elements/reigns/package.json [hunk 1]
```diff
@@ -41,8 +43,10 @@
     "@types/textfit": "^2.4.2",
     "@vitejs/plugin-react": "^1.3.0",
     "babel-jest": "^28.1.0",
+    "canvas": "^2.9.3",
```
Comment (maxired) [inline elements/reigns/package.json:46]: required ? I can't fint it being imported. maybe this is a required peer deps ?

File: elements/reigns/src/AnswerArea.tsx [hunk 2]
```diff
@@ -1,16 +1,35 @@
...
+      <div className="answer__text-outer-container">
+        <div className="answer__text-inner-container">
+          <div ref={ref} id={`${answer}-answer-background`}>
+            <div className="answer__text" style={{ zIndex: 10 }}>
+              {text}
+            </div>
+          </div>
+        </div>
+        <div className="answer__votes-missing" data-testid={`${answer}-votes-missing`}>
+          {!!votesMissing && votesMissing < 4 && <span>{votesMissing} votes missing</span>}
```
Comment (maxired) [inline elements/reigns/src/AnswerArea.tsx:31]: no singular form ?

File: elements/reigns/src/AnswerArea.tsx [hunk 3]
```diff
@@ -1,16 +1,35 @@
...
+  const ref = useRoundedRectangleProgress(progress, color);
+
   return (
     <div className="answer">
-      <div className="answer__text-container">
-        <div className="answer__text">{text}</div>
+      <div className="answer__text-outer-container">
+        <div className="answer__text-inner-container">
+          <div ref={ref} id={`${answer}-answer-background`}>
+            <div className="answer__text" style={{ zIndex: 10 }}>
```
Comment (maxired) [inline elements/reigns/src/AnswerArea.tsx:25]: why inline style ?

File: elements/reigns/src/Game.test.tsx [hunk 4]
```diff
@@ -90,4 +96,75 @@ describe("Game", () => {
...
+      mockSdk({ remoteParticipants });
+      
+      const { getByTestId } = renderGame({ store });
+      persistParticipantVote("1", "Yes");
+      persistParticipantVote("2", "Yes");
+      act(() => {
+        store.dispatch(updateVote());
+      });
+
+      expect(getByTestId("yes-votes-missing")).toHaveTextContent("");
```
Comment (maxired) [inline elements/reigns/src/Game.test.tsx:167]: can we use the parent instead ?
maybe it does not even make sens to render this empty div

File: elements/reigns/src/features/voting/calculateAnswer.test.ts [hunk 5]
```diff
@@ -3,65 +3,124 @@ import { mockSdk } from "../../mocks";
...
-      ],
+  });
+  describe("answer", () => {
+    beforeEach(() => {
+      mockSdk({ remoteParticipants: [remoteParticipant1, remoteParticipant2] });
     });
-    const sdk = getSdk();
-    // put participants inside the element
-    [sdk.localParticipant.id, "remote1", "remote2"].forEach((id) => {
-      sdk.storage.realtime.set(PARTICIPANT_INSIDE_TABLE, id, true);
```
Comment (maxired) [inline elements/reigns/src/features/voting/calculateAnswer.test.ts:19]: was this useless ?

Comment (maxired) [reply elements/reigns/src/features/voting/calculateAnswer.test.ts:19]: done in https://github.com/fres-co/fresco-community/pull/96/files#diff-2fb558bb7b7af48a3d695b4cc12f4768aafddfd208123783fd57118649472f39R52

File: elements/reigns/src/features/voting/calculateAnswer.ts [hunk 6]
```diff
@@ -25,24 +37,37 @@ export const calculateAnswer = (): CalculatedAnswer => {
...
+      votesNeededForAbove50Percent - results.answerNo,
+      0
+    ),
+    yesVotesMissing: Math.max(
+      votesNeededForAbove50Percent - results.answerYes,
+      0
+    ),
+  };
+
+  console.log("progress", results, progress);
```
Comment (maxired) [inline elements/reigns/src/features/voting/calculateAnswer.ts:55]: to clean ?

File: elements/reigns/src/features/voting/useRoundedRectangleProgress.tsx [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+
+export const useRoundedRectangleProgress = (
+  progress: number,
+  color: string
+) => {
+  const ref = useRef<HTMLDivElement>(null);
+  const [pixi, setPixi] = useState<Application>();
+  const currentPositionRef = useRef<number>(0);
+
+  useEffect(() => {
```
Comment (maxired) [inline elements/reigns/src/features/voting/useRoundedRectangleProgress.tsx:19]: not sure if it's worth creating this inside an effect (an so triggering an extra render)

**Review by maxired — APPROVED** [review]
some comments, no major blocker.

I am not sure about the tweeting when teleporting, but I know this will be rework pretty soon, so I guess that's ok

File: elements/reigns/package.json (hunk 1 above)
Comment (staff0rd) [reply elements/reigns/package.json:46]: it's nice to have. without it, `Game.test.tsx` does a `console.error` with the following:
```
Error: Not implemented: HTMLCanvasElement.prototype.getContext (without installing the canvas npm package)
```
It doesn't break the test, but i'd prefer not to have the spam.

File: elements/reigns/src/AnswerArea.tsx (hunk 2 above)
Comment (staff0rd) [reply elements/reigns/src/AnswerArea.tsx:31]: thx, updated.

File: elements/reigns/src/AnswerArea.tsx (hunk 3 above)
Comment (staff0rd) [reply elements/reigns/src/AnswerArea.tsx:25]: thx, was redundant - removed.

File: elements/reigns/src/Game.test.tsx (hunk 4 above)
Comment (staff0rd) [reply elements/reigns/src/Game.test.tsx:167]: the empty div stays so it's sibling element doesn't resize

File: elements/reigns/src/features/voting/useRoundedRectangleProgress.tsx (hunk 7 above)
Comment (staff0rd) [reply elements/reigns/src/features/voting/useRoundedRectangleProgress.tsx:19]: yep, needs ref.

**Comment by staff0rd** [comment]
> not sure about the tweeting when teleporting

to start with it was unexpected, but then I felt afterwards like it was giving it some more life to an otherwise static experience.

**Pull request merged**